### PR TITLE
[RCA Repair Item]Retried deletes should ignore 404 errors

### PR DESCRIPTION
**Pull request by pranavmalik-msft**
**Description:**
This pull request aims to resolve the reported issue related to Azcopy, where it erroneously reports failures in cases of a 404 error, despite a successful background delete operation. The proposed fix involves disregarding 404 errors and treating them as successful operations.

**Changes:**

- Implemented a comprehensive fix to address the observed behavior, ensuring Azcopy operates correctly even when a 404 error is reported.
- Added validation checks to enhance the overall reliability of Azcopy.
- Added Testing Support for this functionality.

### CODE REVIEW

File: ste/sender-appendBlob_test.go [hunk 1]
```diff
@@ -108,3 +109,52 @@ func Test500FollowedBy412Logic(t *testing.T) {
 	a.Nil(err)
 	a.Empty(errString)
 }
+
+// This function tests the scenario where we return a transfer success even when we receive a 404 response, indicating a resource not found error.
+// In this test, we create a container on an HNS enabled account but do not create any blob. This is done to simulate the 404 scenario when attempting to delete a non-existent blob.
+// The deletion operation won't find the blob to delete, resulting in a 404 error, and thus returning a transfer success.
+func Test404DeleteSuccessLogic(t *testing.T) {
```
Comment (nakulkar-msft) [inline ste/sender-appendBlob_test.go:116]: Move this to a different file.

**Review by nakulkar-msft — APPROVED** [review]

File: ste/sender-deleteHNSBlob_test.go [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+	jptm := &testJobPartTransferManager{
+		info: to.Ptr(TransferInfo{
+			Source:       sasURL,
+			SrcContainer: cName,
+			SrcFilePath:  sourceName,
+		}),
+		fromTo: common.EFromTo.BlobFSTrash(),
+	}
+	jptm.SetStatus(common.ETransferStatus.Started())
+	doDeleteHNSResource(jptm)
```
Comment (vibhansa-msft) [inline ste/sender-deleteHNSBlob_test.go:60]: Where are we simulating 404 here?

Comment (pranavmalik-msft) [reply ste/sender-deleteHNSBlob_test.go:60]: to verify the correctness of our new logic

File: ste/xfer-deleteBlobFS.go [hunk 3]
```diff
@@ -51,7 +54,23 @@ func doDeleteHNSResource(jptm IJobPartTransferMgr) {
...
+			var respErr *azcore.ResponseError
+			if errors.As(err, &respErr) {
+				if respErr.StatusCode == http.StatusNotFound {
+					// if the delete failed with err 404, i.e resource not found, then mark the transfer as success.
+					status = common.ETransferStatus.Success()
+				}
+				// If the status code was 403, it means there was an authentication error and we exit.
+				// User can resume the job if completely ordered with a new sas.
+				if respErr.StatusCode == http.StatusForbidden {
+					errMsg := fmt.Sprintf("Authentication Failed. The SAS is not correct or expired or does not have the correct permission %s", err.Error())
```
Comment (vibhansa-msft) [inline ste/xfer-deleteBlobFS.go:66]: Should we assume its only SAS, what about expired OAuth tokens here. Error message shall be generic if any other auth mode also fails with similar error here.

Comment (pranavmalik-msft) [reply ste/xfer-deleteBlobFS.go:66]: In numerous instances, particularly when addressing the deletion of a blob or file, our error handling strategy tends to focus on handling 404 and 403 errors exclusively. Currently, the error message remains intentionally generic to accommodate various scenarios apart from failed snapshots in some instances.

**Review by vibhansa-msft — APPROVED** [review]

**Review by siminsavani-msft — DISMISSED** [review]
thanks for including a clear description @pranavmalik-msft !

**Comment by siminsavani-msft** [comment]
Don't forget to change the base branch to main btw!

**Review by siminsavani-msft — APPROVED** [review]

**Review by gapra-msft — APPROVED** [review]

**Pull request merged**